### PR TITLE
SNOW-645253 Handle binding named parameters

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description
Fixes https://snowflakecomputing.atlassian.net/browse/SNOW-645253 / https://github.com/snowflakedb/gosnowflake/issues/644
Binding does not work for two cases (date/time) - but it doesn't work for any times of binding, not just for named parameters. Need to investigate it further, will provide solution in another PR.

### Checklist
- [ ] Code compiles correctly
- [ ] Run ``make fmt`` to fix inconsistent formats
- [ ] Run ``make lint`` to get lint errors and fix all of them
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: bindings_test.go [hunk 1]
```diff
@@ -930,3 +930,137 @@ func TestFunctionParameters(t *testing.T) {
...
+			transform: func(v interface{}) interface{} { return v },
+		},
+		{
+			desc:      ":1",
+			insert:    "INSERT INTO BINDING_MODES VALUES(:1)",
+			query:     "SELECT * FROM BINDING_MODES WHERE param1 = :1",
+			transform: func(v interface{}) interface{} { return v },
+		},
+		{
+			desc:      ":id",
```
Comment (sfc-gh-dprzybysz) [inline bindings_test.go:972]: should be `:param`?

File: bindings_test.go [hunk 2]
```diff
@@ -930,3 +930,137 @@ func TestFunctionParameters(t *testing.T) {
...
+			query:     "SELECT * FROM BINDING_MODES WHERE param1 = :param",
+			transform: func(v interface{}) interface{} { return sql.Named("param", v) },
+		},
+	}
+
+	runTests(t, dsn, func(dbt *DBTest) {
+		for _, tc := range testcases {
+			for _, bindingMode := range bindingModes {
+				t.Run(tc.testDesc+" "+bindingMode.desc, func(t *testing.T) {
+					query := fmt.Sprintf(`CREATE OR REPLACE TABLE BINDING_MODES(param1 %v)`, tc.paramType)
```
Comment (sfc-gh-dprzybysz) [inline bindings_test.go:983]: who drops that table after tests?

File: bindings_test.go [hunk 3]
```diff
@@ -930,3 +930,137 @@ func TestFunctionParameters(t *testing.T) {
...
+					if !rows.Next() {
+						t.Fatal("Expected to return a row")
+					}
+				})
+			}
+		}
+	})
+}
+
+func TestVariousBindingModesForNullValues(t *testing.T) {
```
Comment (sfc-gh-dprzybysz) [inline bindings_test.go:1001]: what's the difference between these suite and the one above?

File: bindings_test.go [hunk 4]
```diff
@@ -930,3 +930,137 @@ func TestFunctionParameters(t *testing.T) {
...
+			query:     "SELECT * FROM BINDING_MODES WHERE param1 IS NULL",
+			transform: func(v interface{}) interface{} { return sql.Named("param", v) },
+		},
+	}
+
+	runTests(t, dsn, func(dbt *DBTest) {
+		for _, tc := range testcases {
+			for _, bindingMode := range bindingModes {
+				t.Run(tc.testDesc+" "+bindingMode.desc, func(t *testing.T) {
+					query := fmt.Sprintf(`CREATE OR REPLACE TABLE BINDING_MODES(param1 %v)`, tc.paramType)
```
Comment (sfc-gh-dprzybysz) [inline bindings_test.go:1050]: who drops that table?

File: bindings_test.go [hunk 5]
```diff
@@ -930,3 +930,137 @@ func TestFunctionParameters(t *testing.T) {
...
+			transform: func(v interface{}) interface{} { return v },
+		},
+		{
+			desc:      ":1",
+			insert:    "INSERT INTO BINDING_MODES VALUES(:1)",
+			query:     "SELECT * FROM BINDING_MODES WHERE param1 IS NULL",
+			transform: func(v interface{}) interface{} { return v },
+		},
+		{
+			desc:      ":id",
```
Comment (sfc-gh-dprzybysz) [inline bindings_test.go:1039]: should be `:param`?

File: bindings_test.go (hunk 1 above)
Comment (sfc-gh-pfus) [reply bindings_test.go]: Yes it should.

File: bindings_test.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply bindings_test.go:983]: The whole schema is dropped when tests are finished.

File: bindings_test.go (hunk 3 above)
Comment (sfc-gh-pfus) [reply bindings_test.go:1001]: There are nulls tested. Nulls are handled in a different manner in Go which supports nulls only for pointers. Also, the query is also different.

File: bindings_test.go (hunk 5 above)
Comment (sfc-gh-pfus) [reply bindings_test.go]: Yes, it should.

File: bindings_test.go (hunk 4 above)
Comment (sfc-gh-pfus) [reply bindings_test.go:1050]: Same as above.

File: bindings_test.go (hunk 3 above)
Comment (sfc-gh-dprzybysz) [reply bindings_test.go:1001]: In the new test there is a lot of duplication - maybe you can extract one parameter more and then put those tests in one suite?

File: bindings_test.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply bindings_test.go:983]: No, in developer machine it's the same. See [teardown](https://github.com/snowflakedb/gosnowflake/blob/master/driver_test.go#L134) invoked after [all tests](https://github.com/snowflakedb/gosnowflake/blob/master/driver_test.go#L158).

File: bindings_test.go (hunk 3 above)
Comment (sfc-gh-pfus) [reply bindings_test.go:1001]: I tried - but it's less readable. You have to change query between `= ?`/`= :1`/`= :1` and `is null`. IMHO it wasn't worth it.

Comment (sfc-gh-dheyman-1) [reply bindings_test.go:1001]: I think I would consolidate these. Seems like theres no point in maintaining the same test running logic twice. Also It seems like we would only add some scenarios' config so readability wouldn't suffer that much

**Review by sfc-gh-dprzybysz — APPROVED** [review]

**Review by sfc-gh-pbulawa — APPROVED** [review]

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM

**Pull request merged**